velox/type/Filter.cpp

```diff
@@ -98,6 +97,626 @@ std::string Filter::toString() const {
+}
+
+folly::dynamic AlwaysTrue::serialize() const {
+  return Filter::serializeBase("AlwaysTrue");
+}
+
+FilterPtr AlwaysTrue::create(const folly::dynamic& /*obj*/) {
+  return std::make_unique<AlwaysTrue>();
+}
+
+folly::dynamic IsNull::serialize() const {
+  return Filter::serializeBase("IsNull");
+}
+
+FilterPtr IsNull::create(const folly::dynamic& /*obj*/) {
+  return std::make_unique<IsNull>();
+}
+
+folly::dynamic IsNotNull::serialize() const {
+  return Filter::serializeBase("IsNotNull");
+}
+
+FilterPtr IsNotNull::create(const folly::dynamic& /*obj*/) {
+  return std::make_unique<IsNotNull>();
+}
+
+folly::dynamic BoolValue::serialize() const {
+  auto obj = Filter::serializeBase("BoolValue");
+  obj["value"] = value_;
+  return obj;
+}
+
+FilterPtr BoolValue::create(const folly::dynamic& obj) {
+  auto value = obj["value"].asBool();
+  auto nullAllowed = deserializeNullAllowed(obj);
+  return std::make_unique<BoolValue>(value, nullAllowed);
+}
+
+bool BoolValue::testingEquals(const Filter& other) const {
+  auto otherBoolValue = dynamic_cast<const BoolValue*>(&other);
+  return otherBoolValue != nullptr && Filter::testingBaseEquals(other) &&
+      (value_ == otherBoolValue->value_);
+}
+
+folly::dynamic BigintRange::serialize() const {
+  auto obj = Filter::serializeBase("BigintRange");
+  obj["lower"] = lower_;
+  obj["upper"] = upper_;
+  return obj;
+}
+
+FilterPtr BigintRange::create(const folly::dynamic& obj) {
+  auto lower = obj["lower"].asInt();
+  auto upper = obj["upper"].asInt();
+  auto nullAllowed = deserializeNullAllowed(obj);
+  return std::make_unique<BigintRange>(lower, upper, nullAllowed);
+}
+
+bool BigintRange::testingEquals(const Filter& other) const {
+  auto otherBigintRange = dynamic_cast<const BigintRange*>(&other);
+  return otherBigintRange != nullptr && Filter::testingBaseEquals(other) &&
+      (lower_ == otherBigintRange->lower_) &&
+      (upper_ == otherBigintRange->upper_);
+}
+
+folly::dynamic NegatedBigintRange::serialize() const {
+  auto obj = Filter::serializeBase("NegatedBigintRange");
+  obj["lower"] = nonNegated_->lower();
+  obj["upper"] = nonNegated_->upper();
+  return obj;
+}
+
+FilterPtr NegatedBigintRange::create(const folly::dynamic& obj) {
+  auto lower = obj["lower"].asInt();
+  auto upper = obj["upper"].asInt();
+  auto nullAllowed = deserializeNullAllowed(obj);
+  return std::make_unique<NegatedBigintRange>(lower, upper, nullAllowed);
+}
+
+bool NegatedBigintRange::testingEquals(const Filter& other) const {
+  auto otherNegatedBigintRange =
+      dynamic_cast<const NegatedBigintRange*>(&other);
+  return otherNegatedBigintRange != nullptr &&
+      Filter::testingBaseEquals(other) &&
+      nonNegated_->testingEquals(*(otherNegatedBigintRange->nonNegated_));
+}
+
+folly::dynamic HugeintRange::serialize() const {
+  auto obj = Filter::serializeBase("HugeintRange");
+  auto size = sizeof(int128_t);
+  char lower[size];
+  HugeInt::serialize(lower_, lower);
```

Comment:
A quick question, are we writing these serialization as JSON?  If yes, some of these strings are not printable

Reply:
No, HugeInt::serialize use memcpy to copy the int128 value's bytes into a char array.

```C++
static FOLLY_ALWAYS_INLINE int128_t deserialize(const char* serializedData) {
    int128_t value;
    memcpy(&value, serializedData, sizeof(int128_t));
    return value;
  }
```

Reply:
I've updated `HugeintRange` SerDe methods. Serialize the int128 value using to_string and deserialize it using the `HugeInt::stringToInt128` (a new added method in this pr ).
```C++
// Ser
obj["lower"] = std::to_string(lower_);
obj["upper"] = std::to_string(upper_);
// De
auto lower = HugeInt::stringToInt128(obj["lower"].asString());
auto upper = HugeInt::stringToInt128(obj["upper"].asString());
```
Now the output of toJson(hugeIntRangeObj) would be as follow,
`{"upper":"-555586388888891223462895","kind":"kHugeintRange","lower":"-111914766488860230291735206771233412079","nullAllowed":true,"name":"HugeintRange"}`

Reply:
As discussed in Slack, we can use 2 int64_t for lower and upper, or use one string for the whole int128